### ISSUE TITLE
Public-Signed Certificate Validation Error

### ISSUE THREAD

**Issue by KenHolloway** [FreeRDP/FreeRDP#4768]
I recently upgraded to openSUSE Leap 15.0 on my laptop, and now I am getting a "certificate could not be verified" warning when connecting to servers that are using a public-signed certificate.

I redownloaded and recompiled on both openSUSE Leap 42.3 and openSUSE Leap 15.0 today - it works on 42.3, but not on the newer 15.0.

I am attaching the output from a strace for both operating systems. It looks to me like both are checking for /var/lib/ca-certificates/openssl/4bfab552.0 (which exists, and is the correct CA) - but on 15.0, it never opens that file?

[openSUSE-Leap-42.3-Works.TXT](https://github.com/FreeRDP/FreeRDP/files/2229158/openSUSE-Leap-42.3-Works.TXT)
[openSUSE-Leap-15-Error.TXT](https://github.com/FreeRDP/FreeRDP/files/2229159/openSUSE-Leap-15-Error.TXT)

**Comment by akallabeth** [comment]
@KenHolloway yes, we currently only implement certificate fingerprint matching, no check for authorities.

**Comment by KenHolloway** [comment]
I am not sure what the exact terminology is, but when I am running under openSUSE 13.1 or openSUSE 42.3, connecting to a server that is using a public-signed certificate just connects; when I try the exact same thing under openSUSE Leap 15.0, I get an error.

I was not trying to ask for a new feature - just help figuring out why it behaves differently under the different versions of openSUSE, and how to make it behave "right" under openSUSE Leap 15.0 (the newest version, regardless of the version numbers)

Thanks,
Ken

**Comment by akallabeth** [comment]
Ok, did you accept keys previously? /cert-tofu does that on first connect

**Comment by KenHolloway** [comment]
No, I had not accepted the keys - known_hosts2 is empty on my openSUSE 42.3 machine that I am testing from, and I have never used /cert-tofu.

In the output from strace, I see that on both operating systems it is finding that the CA certificate exists:

stat("/var/lib/ca-certificates/openssl/4bfab552.0", {st_mode=S_IFREG|0444, st_size=1492, ...}) = 0

Under openSUSE Leap 42.3, it then opens the CA:

open("/var/lib/ca-certificates/openssl/4bfab552.0", O_RDONLY|O_CLOEXEC) = 13

But, under openSUSE Leap 15.0, I see no attempt to open the CA.

**Comment by akallabeth** [comment]
@KenHolloway Just checked, are there some OpenSSL related settings changed? We're verifying the certificate in `x509_verify_certificate` which is a wrapper creating a `X509_STORE` with additional lookup paths.

**Comment by KenHolloway** [comment]
I have not ever made any OpenSSL configuration changes.

I just compared /etc/sll/openssl.cnf on the two systems. The only differences I see are that openSUSE Leap 42.3 had:

basicConstraints = CA:true
digests         = md5, sha1

and openSUSE Leap 15.0 has:

basicConstraints = critical,CA:true
signer_digest  = sha256
digests     = sha1, sha256, sha384, sha512


I tried copying openssl.cnf from the 15.0 system to the 42.3 system, but it made no difference.

**Comment by KenHolloway** [comment]
I am not sure how to tell. This certificate is from 27 December 2016. When I look at it in a browser (same certificate is used for IIS and RDP), it says Certificate Signature Algorithm is SHA-256.

**Comment by akallabeth** [comment]
@KenHolloway ok, so the cert is sha256. Could you check the ca signed with sha256 too? Or, inverse test does it work if you add sha1 to signer_digest? (Please remove again after the test)

**Comment by KenHolloway** [comment]
Both the Intermediate and Root certificates are also SHA-256.

I copied the opensssl.cnf from my 42.3 computer to my 15.0 computer - no change. (strace still shows that it finds that /var/lib/ca-certificates/openssl/4bfab552.0 exists, but it never tries to open in. Under 42.3, it find that it exists, then opens it.)

**Comment by KenHolloway** [comment]
Does FreeRDP just call OpenSSL to do the certificate checking? Is this a problem with OpenSSL instead of FreeRDP?

**Comment by akallabeth** [comment]
@KenHolloway Roughly yes, but `x509_verify_certificate` does some argument conversion/preparation before calling openssl

**Comment by akallabeth** [comment]
@KenHolloway You don't by chance have a link to a proper documentation of that? The API docs of OpenSSL are light on the workings of this one.

**Comment by KenHolloway** [comment]
Unfortunately, no. I too felt the documentation was a bit sparse. I found that code in a posting about a similar problem. I also found another posting that said:

store = X509_STORE_new();
verify_params = X509_Store_get0_param(store);
purpose = X509_PURPOSE_get_by_sname("sslclient");
X509_VERIFY_PARAM_set_purpose(verify_params, purpose)
...
PKCS7_verify(..., store, ...);

I don't know which way is "better". I also did not save the URL to either post....

**Comment by KenHolloway** [comment]
This change did not fix the issue for me. I am still getting:

```
The above X.509 certificate could not be verified, possibly because you do not have
the CA certificate in your certificate store, or the certificate has expired.
```

I added a call to X509_STORE_CTX_get_error - it is returning error 28:

```
X509_V_ERR_CERT_REJECTED: certificate rejected
    the root CA is marked to reject the specified purpose.
```

**Comment by akallabeth** [comment]
@KenHolloway if you change to purpose server does that work?

**Comment by KenHolloway** [comment]
I tried X509_PURPOSE_SSL_SERVER, and X509_PURPOSE_SSL_CLIENT - it did not make any difference.

Do we need to "X509_STORE_CTX_set0_param(csc,verify_param)"? (It seems like, after getting that param, we might need to set it back.) I tried adding that, but then got a different error (22):

```
X509_V_ERR_CERT_CHAIN_TOO_LONG: certificate chain too long
    the certificate chain length is greater than the supplied maximum depth. Unused.
```
